File: src/main/java/no/fintlabs/instance/gateway/InstanceController.java
package no.fintlabs.instance.gateway;

import lombok.extern.slf4j.Slf4j;
import no.fintlabs.gateway.instance.InstanceProcessor;
import no.fintlabs.gateway.instance.kafka.ArchiveCaseIdRequestService;
import no.fintlabs.instance.gateway.model.Status;
import no.fintlabs.instance.gateway.model.vigo.IncomingInstance;
import no.fintlabs.resourceserver.security.client.sourceapplication.SourceApplicationAuthorizationService;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;
import reactor.core.publisher.Mono;

import static no.fintlabs.resourceserver.UrlPaths.EXTERNAL_API;

@Slf4j
@RestController
@RequestMapping(EXTERNAL_API + "/vigo/instances")
public class InstanceController {

    private final InstanceProcessor<IncomingInstance> instanceProcessor;

    private final ArchiveCaseIdRequestService archiveCaseIdRequestService;

    private final SourceApplicationAuthorizationService sourceApplicationAuthorizationService;

    public InstanceController(
            InstanceProcessor<IncomingInstance> instanceProcessor,
            ArchiveCaseIdRequestService archiveCaseIdRequestService,
            SourceApplicationAuthorizationService sourceApplicationAuthorizationService
    ) {
        this.instanceProcessor = instanceProcessor;
        this.archiveCaseIdRequestService = archiveCaseIdRequestService;
        this.sourceApplicationAuthorizationService = sourceApplicationAuthorizationService;
    }

    @PostMapping("instance")
    public Mono<ResponseEntity<?>> postIncomingInstance(
            @RequestBody IncomingInstance incomingInstance,
            @AuthenticationPrincipal Mono<Authentication> authenticationMono
    ) {
        log.debug("Incoming instance: {}", incomingInstance.getInstansId());

        return authenticationMono.flatMap(
                authentication -> instanceProcessor.processInstance(
                        authentication,
                        incomingInstance
                )
        );
    }

    @GetMapping("status/{instanceId}")
    public Mono<ResponseEntity<Status>> getInstanceStatus(
            @AuthenticationPrincipal Mono<Authentication> authenticationMono,
            @PathVariable String instanceId
    ) {
        return authenticationMono.map(authentication ->
                {
                    Long applicationId = sourceApplicationAuthorizationService.getSourceApplicationId(authentication);

                    log.debug("Get status for instance: {} in sourceApplication: {}", instanceId, applicationId);

                    return archiveCaseIdRequestService.getArchiveCaseId(applicationId, instanceId)
                            .map(caseId -> ResponseEntity.ok(Status.builder()
                                            .instansId(instanceId)
                                            .destinasjonsId(caseId)
                                            .status("Instans godtatt av destinasjon").build()
                                    )
                            )
                            .orElse(ResponseEntity
                                    .badRequest()
                                    .body(Status.builder()
                                            .instansId(instanceId)
                                            .status("Ukjent status").build()
                                    )
                            );
                }
        );
    }

}

File: src/main/kotlin/no/fintlabs/instance/gateway/InstanceController.kt
package no.fintlabs.instance.gateway

import no.fintlabs.gateway.instance.InstanceProcessor
import no.fintlabs.gateway.instance.kafka.ArchiveCaseIdRequestService
import no.fintlabs.instance.gateway.model.Status
import no.fintlabs.instance.gateway.model.vigo.IncomingInstance
import no.fintlabs.resourceserver.UrlPaths.EXTERNAL_API
import no.fintlabs.resourceserver.security.client.sourceapplication.SourceApplicationAuthorizationService
import org.slf4j.LoggerFactory
import org.springframework.http.ResponseEntity
import org.springframework.security.core.Authentication
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.web.bind.annotation.*
import reactor.core.publisher.Mono

@RestController
@RequestMapping("$EXTERNAL_API/vigo/instances")
class InstanceController(
    private val instanceProcessor: InstanceProcessor<IncomingInstance>,
    private val archiveCaseIdRequestService: ArchiveCaseIdRequestService,
    private val sourceApplicationAuthorizationService: SourceApplicationAuthorizationService
) {
    private val log = LoggerFactory.getLogger(InstanceController::class.java)

    @PostMapping("instance")
    fun postIncomingInstance(
        @RequestBody incomingInstance: IncomingInstance,
        @AuthenticationPrincipal authenticationMono: Mono<Authentication>
    ): Mono<ResponseEntity<*>> {
        log.debug("Incoming instance: {}", incomingInstance.instansId)

        return authenticationMono.flatMap { authentication ->
            instanceProcessor.processInstance(
                authentication,
                incomingInstance
            )
        }
    }
}
